dkg/keycast.go

```diff
@@ -0,0 +1,246 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package dkg
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	"github.com/coinbase/kryptology/pkg/signatures/bls/bls_sig"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+	"github.com/obolnetwork/charon/cluster"
+	"github.com/obolnetwork/charon/tbls"
+)
+
+// transport provides secure transport abstraction to keycast.
+type transport interface {
+	// ServeShares registers a function that serves node share messages until the context is closed.
+	ServeShares(context.Context, func(nodeIdx int) (msg []byte, err error))
+	// GetShares returns the shares served by a dealer or a permanent error.
+	GetShares(ctx context.Context, nodeIdx int) ([]byte, error)
+}
+
+// share is the co-validator public key, tbls verifiers, and private key share.
+// Each node in the cluster will receive one for each distributed validator.
+type share struct {
+	PubKey   *bls_sig.PublicKey
+	Verifier *sharing.FeldmanVerifier
+	Share    *bls_sig.SecretKeyShare
+}
+
+// shareMsg is the share message wire format sent by the dealer.
+type shareMsg struct {
+	PubKey    []byte
+	Verifiers [][]byte
+	Share     []byte
+}
+
+func runKeyCast(ctx context.Context, def cluster.Definition, tx transport, nodeIdx int, random io.Reader) ([]share, error) {
```

Comment:
nit: can we rename `tx` to simply `t` or `tp` coz my brain simply converts `tx` to `transaction`

Reply:
i would also suggest to rename tx to something else(t, tp), since tx in eth terms can be easily misunderstood to transaction.